Promote SidecarContainers feature to GA

gjkim42 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind feature

#### What this PR does / why we need it:

This PR promotes the SidecarContainers feature to the GA.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
The `SidecarContainers` feature has graduated to GA. 'SidecarContainers' feature gate was locked to default value and will be removed in v1.36. If you were setting this feature gate explicitly, please remove it now.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/753-sidecar-containers/README.md
```

gjkim42 commented:
/cc @matthyx @SergeyKanzhelev 
/priority important-soon

gjkim42 commented on pkg/kubelet/kuberuntime/kuberuntime_manager.go:
```diff
@@ -934,6 +935,8 @@ func (m *kubeGenericRuntimeManager) computePodActions(ctx context.Context, pod *
 		if len(containersToStart) == 0 {
 			hasInitialized := false
+			// TODO: Remove this code path as logically it is the subset of the next
+			// code path.
```
xref: https://github.com/kubernetes/kubernetes/pull/127162#discussion_r1923868666

toVersus replied:
I'm not entirely sure about the intent behind keeping this code path. When will it be possible to remove it completely?

SergeyKanzhelev replied:
we will need to introduce a separate feature gate so we switch logic safely. Keeping two code path doesn't really makes sense long term, but short term we want to be extra cautious

SergeyKanzhelev commented:
Any reason to keep the WIP label?

gjkim42 commented:
> Any reason to keep the WIP label?


@SergeyKanzhelev 

https://github.com/kubernetes/kubernetes/pull/129731#discussion_r1923869901

only to address this comment.

We'll leave the TODO comment here and add a new PR to create a temporary feature gate to remove the feature gate in the next release, right?

gjkim42 commented:
cleaned up the commits and removed the WIP label from the PR.

/assign @SergeyKanzhelev 
Could you PTAL?

SergeyKanzhelev commented on pkg/apis/core/validation/validation.go:
```diff
@@ -5616,19 +5616,18 @@ func ValidatePodResize(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
 	}
 
 	// Do not allow removing resource requests/limits on resize.
-	if utilfeature.DefaultFeatureGate.Enabled(features.SidecarContainers) {
```
Since we need to support an `emulated_version`, we need to keep this code for control plane. See KEP: https://github.com/kubernetes/enhancements/issues/4330 Let's still delete it from the kubelet code.

gjkim42 replied:
done

SergeyKanzhelev commented on test/integration/scheduler/scoring/priorities_test.go:
```diff
@@ -253,7 +253,6 @@ func TestNodeResourcesScoring(t *testing.T) {
 
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.SidecarContainers, true)
```
we will need to rewrite it to use emulated_version instead to test this logic

gjkim42 replied:
I guess the emulated version is only needed when disabling the feature gate.
Updated the tests disabling the feature to use the emulated version.

gjkim42 committed:
Promote SidecarContainers feature to GA

gjkim42 committed:
Leave TODOs in pkg/kubelet/kuberuntime for later removal

This leaves TODOs in pkg/kubelet/kuberuntime to remove these redundant
code paths later, since they are supposed to be a subset of the new code
paths.

SergeyKanzhelev reviewed (approved):
/lgtm
/approve

hold for KEP GA approval: https://github.com/kubernetes/enhancements/pull/5081

/hold

SergeyKanzhelev commented:
/unhold

/assign @janetkuo @msau42

janetkuo commented:
@gjkim42 in the release note `'SidecarContainers' feature was locked to default value and will be removed in v1.36.` Would you change it from "feature" to "feature gate" to make it more clear that the feature isn't going to be removed, and just the feature gate will be?

janetkuo commented:
/approve

msau42 commented:
/approve

for API
